<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta charset="utf-8">
    <title>BastyGenerateFasta</title>

    <!-- Latest compiled and minified CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">

    <!-- Optional theme -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap-theme.min.css" integrity="sha384-rHyoN1iRsVXV4nD0JutlnGaslCJuC7uwjduW9SVrLvRYooPp2bWYgmgJQIXwl/Sp" crossorigin="anonymous">

    <!-- jQuery (necessary for Bootstrap's JavaScript plugins) -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>

    <!-- Latest compiled and minified JavaScript -->
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link href="css/docs.css" rel="stylesheet">

  </head>

  <body>
  <nav class="navbar navbar-default navbar-fixed-top">
    <div class="container-fluid">
      <div class="navbar-header">
        <div class="navbar-header">
          <a class="navbar-brand" href="#">BastyGenerateFasta</a>
          <ul class="nav navbar-nav navbar-right">
            <li class="navbar-text" id="Version">
              Version</li>
          </ul>
        </div>
      </div>



    </div>

  </nav>
  <div class="container">

    <div class="row">
      <div class="col-sm-3 col-md-2 sidebar" >

        <ul class="nav nav-sidebar nav-list">
          <li><b>Index</b></li>
          
          <li><a href="#description"><span>Description</span></a></li>
          
          <li><a href="#installation"><span>Installation</span></a></li>
          
          <li><a href="#manual"><span>Manual</span></a></li>
          
          <li><a href="#about"><span>About</span></a></li>
          
          <li><a href="#contributing"><span>Contributing</span></a></li>
          
          <li><a href="#contact"><span>Contact</span></a></li>
          
        </ul>
        <ul class="nav nav-sidebar nav-list">
          <li><b>Project Links</b></li>

          <li><a href="https://github.com/biopet/bastygeneratefasta">Source code</a></li>
          <li><a href="https://github.com/biopet/bastygeneratefasta/issues">Feedback, feature requests and bug reports</a></li>
          <li><a href="api/index.html">API Documentation</a></li>
        </ul>
        <ul class="nav nav-sidebar nav-list">
          <li><b>Documentation</b></li>
          <li><a href="../index.html">Latest stable version</a></li>
          <li><a href="../develop/index.html">Development version</a></li>
        </ul>
      </div>
      
      <div class="col-sm-9 col-sm-offset-3 col-md-10 col-md-offset-2 main" id="top">

        
        <h1 id="description" class="section">Description</h1>
        <p>This tool generates Fasta files out of variant (SNP) alignments or full alignments (consensus).
        It can be very useful to produce the right input needed for follow up tools,
        for example phylogenetic tree building.</p>
        
        <h1 id="installation" class="section">Installation</h1>
        <p>BastyGenerateFasta requires Java 8 to be installed on your device. Download Java 8
        <a href="http://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html">here</a>
        or install via your distribution&#39;s package manager.</p>
        <p>Download the latest version of BastyGenerateFasta <a href="https://github.com/biopet/bastygeneratefasta/releases/">here</a>.
        To generate the usage run:</p>
        <pre><code>java -jar BastyGenerateFasta --version.jar --help</code></pre>
        
        <h1 id="manual" class="section">Manual</h1>
        <p>BastyGenerateFasta has three modes:</p>
        <ul>
          <li>outputVariants: This mode outputs all the variants in fasta format.</li>
          <li>outputConsensus: This mode outputs consensus sequences from a BAM file.</li>
          <li>outputConsensusVariants: This mode combines the above two.</li>
        </ul>
        
        <h2 id="example" class="section">Example</h2>
        <p>Minimal example for option: <code>--outputVariants</code> (VCF based)</p>
        <pre><code>java -jar &lt;BastyGenerateFasta_jar&gt; \
--inputVcf myVCF.vcf \
--outputName NiceTool \
--outputVariants myVariants.fasta</code></pre>
        <p>Minimal example for option: <code>--outputConsensus</code> (BAM based)</p>
        <pre><code>java -jar &lt;BastyGenerateFasta_jar&gt; \
--bamFile myBam.bam \
--outputName NiceTool \
--outputConsensus myConsensus.fasta \
--reference reference.fa</code></pre>
        <p>Minimal example for option: <code>--outputConsensusVariants</code> (Both)</p>
        <pre><code>java -jar &lt;BastyGenerateFasta_jar&gt; \
--inputVcf myVCF.vcf \
--bamFile myBam.bam \
--outputName NiceTool \
--outputConsensusVariants myConsensusVariants.fasta \
--reference reference.fa</code></pre>
        
        <h2 id="usage" class="section">Usage</h2>
        <p>Usage for BastyGenerateFasta:</p>
        <table class="table">
          <thead>
            <tr>
              <th>Option</th>
              <th>Required</th>
              <th>Can occur multiple times</th>
              <th>Description</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>--log_level, -l</td>
              <td>no</td>
              <td>no</td>
              <td>Level of log information printed. Possible levels: &#39;debug&#39;, &#39;info&#39;, &#39;warn&#39;, &#39;error&#39;</td>
            </tr>
            <tr>
              <td>--help, -h</td>
              <td>no</td>
              <td>no</td>
              <td>Print usage</td>
            </tr>
            <tr>
              <td>--version, -v</td>
              <td>no</td>
              <td>no</td>
              <td>Print version</td>
            </tr>
            <tr>
              <td>--inputVcf, -V</td>
              <td>no</td>
              <td>no</td>
              <td>vcf file, needed for outputVariants and outputConsensusVariants</td>
            </tr>
            <tr>
              <td>--bamFile</td>
              <td>no</td>
              <td>no</td>
              <td>bam file, needed for outputConsensus and outputConsensusVariants</td>
            </tr>
            <tr>
              <td>--outputVariants</td>
              <td>no</td>
              <td>no</td>
              <td>fasta with only variants from vcf file</td>
            </tr>
            <tr>
              <td>--outputConsensus</td>
              <td>no</td>
              <td>no</td>
              <td>Consensus fasta from bam, always reference bases else &#39;N&#39;</td>
            </tr>
            <tr>
              <td>--outputConsensusVariants</td>
              <td>no</td>
              <td>no</td>
              <td>Consensus fasta from bam with variants from vcf file, always reference bases else &#39;N&#39;</td>
            </tr>
            <tr>
              <td>--snpsOnly</td>
              <td>no</td>
              <td>no</td>
              <td>Only use snps from vcf file</td>
            </tr>
            <tr>
              <td>--sampleName</td>
              <td>no</td>
              <td>no</td>
              <td>Sample name in vcf file</td>
            </tr>
            <tr>
              <td>--outputName</td>
              <td>yes</td>
              <td>no</td>
              <td>Output name in fasta file header</td>
            </tr>
            <tr>
              <td>--minAD</td>
              <td>no</td>
              <td>no</td>
              <td>min AD value in vcf file for sample. Defaults to: 8</td>
            </tr>
            <tr>
              <td>--minDepth</td>
              <td>no</td>
              <td>no</td>
              <td>min depth in bam file. Defaults to: 8</td>
            </tr>
            <tr>
              <td>--reference</td>
              <td>no</td>
              <td>no</td>
              <td>Indexed reference fasta file</td>
            </tr>
          </tbody>
        </table>
        
        <h1 id="about" class="section">About</h1>
        <p>BastyGenerateFasta is part of BIOPET tool suite that is developed at LUMC by <a href="http://sasc.lumc.nl/">the SASC team</a>.
        Each tool in the BIOPET tool suite is meant to offer a standalone function that can be used to perform a
        dedicate data analysis task or added as part of <a href="http://biopet-docs.readthedocs.io/en/latest/">BIOPET pipelines</a>.</p>
        <p>All tools in the BIOPET tool suite are <a href="https://www.gnu.org/philosophy/free-sw.html">Free</a> and
        <a href="https://opensource.org/osd">Open Source</a> Software.</p>
        
        <h1 id="contributing" class="section">Contributing</h1>
        <p>The source code of BastyGenerateFasta can be found <a href="https://github.com/biopet/bastygeneratefasta">here</a>.
        We welcome any contributions. Bug reports, feature requests and feedback can be submitted at our
        <a href="https://github.com/biopet/bastygeneratefasta/issues">issue tracker</a>.</p>
        <p>BastyGenerateFasta is build using sbt. Before submitting a pull request, make sure all tests can be passed by
        running <code>sbt test</code> from the project&#39;s root. We recommend using an IDE to work on BastyGenerateFasta. We have had
        good results with <a href="https://www.jetbrains.com/idea/">this IDE</a>.</p>
        
        <h1 id="contact" class="section">Contact</h1>
        <p>
          <!-- Obscure e-mail address for spammers -->
        For any question related to BastyGenerateFasta, please use the
        <a href='https://github.com/biopet/bastygeneratefasta/issues'>github issue tracker</a>
        or contact
         <a href='http://sasc.lumc.nl/'>the SASC team</a> directly at: <a href='&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#115;&#97;&#115;&#99;&#64;&#108;&#117;&#109;&#99;&#46;&#110;&#108;'>
        &#115;&#97;&#115;&#99;&#64;&#108;&#117;&#109;&#99;&#46;&#110;&#108;</a>.
        </p>      

      </div>
    </div>

  </div>
  <script>
    var loc = window.location.pathname;
    var thisDir = loc.substring(0, loc.lastIndexOf('/'));
    var splitLocation = window.location.pathname.split('/');
    var version = splitLocation[splitLocation.length - 2]
    document.getElementById("Version").innerHTML = "Version " + version;
  </script>
</body>

</html>
